TilerServer/TilerWebModule.pas

```diff
@@ -3627,140 +3657,547 @@ function TSliceDiffGeometryICLR.GenerateTileCalc(const aExtent: TExtent; aBitmap
+                      yd := {(1+}widthL*capacityFactor{)}*yn/pointDist;
+                      // todo: wrong rotation direction..
+
+                      polyPoints := [xPrev, yPrev, xCurr, yCurr, xCurr+xd, yCurr+yd, xPrev+xd, yPrev+yd];
+                      polygon := ConstructPolygon(polyPoints);
+
+                      DrawFillPolygon(colorsL, aBitmap, polygon);
+                    end;
+
                     //draw right side
                     if validR then
                     begin
-                      xd := {(1+}widthR*capacityFactor{)}*xn/l;
-                      yd := {(1+}widthR*capacityFactor{)}*yn/l;
-                      polygon[2].X := x-xd;
-                      polygon[2].Y := y-yd;
-                      polygon[3].X := xPrev-xd;
-                      polygon[3].Y := yPrev-yd;
-                      // draw
-                      if colorsR.fillColor<>0  then
+                      xd := {(1+}widthR*capacityFactor{)}*xn/pointDist;
+                      yd := {(1+}widthR*capacityFactor{)}*yn/pointDist;
+
+                      polyPoints := [xPrev, yPrev, xCurr, yCurr, xCurr-xd, yCurr-yd, xPrev-xd, yPrev-yd];
+                      polygon := ConstructPolygon(polyPoints);
+
+                      DrawFillPolygon(colorsR, aBitmap, polygon);
+                    end;
+                  end;
+                end;
+              end;
+            end;
+          end;
+        end;
+        // todo: process ref geometries not in current? or just skip..
+      finally
+        fCurrentSlice.fDataLock.EndRead;
+        fRefSlice.fDataLock.EndRead;
+      end;
+      Result := gtsOk;
+    finally
+      aBitmap.Canvas.EndScene;
+    end;
+  end
+  else Log.WriteLn('TSliceDiffGeometryPolygonLR layer '+fLayer.LayerID.ToString+': no palette defined', llError);
+end;
+
+procedure TSliceDiffGeometryPolygonLR.DrawDefaultPath(path: TPathData; aBitmap: FMX.Graphics.TBitmap);
+begin
+  try
+    aBitmap.Canvas.Stroke.Color := TAlphaColorRec.Black or TAlphaColorRec.Alpha;
+    aBitmap.Canvas.Stroke.Thickness := 1;
+    aBitmap.Canvas.DrawPath(path, 1);
+  finally
+    path.Free;
+  end;
+end;
+
+function TSliceDiffGeometryPolygonLR.CalculateWidth(aActiveValue, aRefValue: Double; var aValidFlag: Boolean): Double;
+begin
+  if not IsNaN(aActiveValue) then
+  begin
+    if not IsNaN(aRefValue) then
+      Result := (aRefValue + aActiveValue) / 2
+    else Result := aActiveValue;
+  end
+  else
+  begin
+    if not IsNaN(aRefValue)
+    then Result := aRefValue
+    else aValidFlag := False;
+  end;
+end;
+
+function TSliceDiffGeometryPolygonLR.GetPaletteColor(aActiveTexture, aRefTexture: Double; var aValidFlag: Boolean): TGeoColors;
+begin
+  if not (IsNaN(aActiveTexture) or IsNaN(aRefTexture))then
+    Result := fPalette.ValueToColors(aActiveTexture-aRefTexture)
+  else aValidFlag := False;
+end;
+
+function TSliceDiffGeometryPolygonLR.ConstructPolygon(aPolyPoints: TArray<Double>): TPolygon;

```

Comment:
I think this is very time consuming way to construct a polygon and performance will suffer greatly. Maybe this should be optimized by re-using the same polygon by passing it by ref ("var" in delphi) and because it is always the same amount of points just code them directly?

Reply:
You are right. Creating another polygon and returning it was not very optimized.
I have passed the polygon by reference and re-used it in the function.
This function (now a procedure, as it does not return anything anymore) was created to remove the redundant code that was initially present for creating polygons in all the three classes for the difference layers.